---
title: "Redeconve manual"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Redeconve}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, warning=FALSE}
library(Redeconve)
```

------------------------------------------------------------------------

## Main function

This part describes how to use the main function `deconvoluting` to perform single-cell deconvolution.

### Useage

The usage of `deconvoluting` is as follows:

```{r}
## nums = deconvoluting(ref, st, cellnames=NULL, genemode, gene.list, var_thresh=0.025, exp_thresh=0.03, hpmode, hp, normalize = T, thre=1e-10, dopar=T, ncores, realtime=F, dir=NULL)
```

It does contains many parameters. Next, I will divide these parameters into several parts by function, and explain them one by one.

### 1. Necessary data for deconvolution

`ref` and `st` are the core data used for deconvolution. They are required as follows:

-   `ref`, the scRNA-seq data served as reference for deconvolution. It is a `Matrix` (or `dgCMatrix`) of unprocessed count-level scRNA-seq data. One row represents one gene and one column represents one cell.
-   `st`, the spatial transcriptomics to be deconvoluted. It is a `Matrix` (or `dgCMatrix`) of unprocessed spatial transcriptomics, represents raw counts in each spot. One row represents one gene and one column represents one spot.

### 2. Mode of gene selection

`genemode`, `gene.list`, `var_thresh` and `exp_thresh` are about how to deal with genes. `genemode` determines the mode of handling genes and `gene.list`, `var_thresh`, `exp_thresh` are associated with specific modes. Redeconve offers 3 alternative modes of dealing with genes:

1.  `default`: Use the intersection of genes in `ref` and `st`, without other treatment.

2.  `customized`: Indicating gene list yourself. Parameter `gene.list` is the list of genes you indicated. Note that only those genes within the intersection of `ref` and `st` would be used.

3.  `filtered`: We will use a built-in function `gene.filter` to screen some genes. This function will first take the intersection of `ref` and `st`, the use two indices, `var_thresh` and `exp_thresh` to filter genes. You can customize these two parameters as well.

-   `var_thresh` considers variance of reference. Genes whose variance across all cells in reference do not reach that threshold will be filtered out. The default value is 0.025.
-   `exp_thresh` considers expression in spatial transcriptomics. Genes whose average count across all spots in spatial transcriptomics is less that this value will be filtered out. The default value is 0.003.

### 3. Mode of determining hyperparameter

The hyperparamter is our key to single-cell resolution (See Methods for details). Here we still offers 3 modes to determine the hyperparameter:

1.  `default`: We will calculate a hyperparameter according to the number of genes and cells in reference (See Methods for details).

2.  `customized`: Indicating the hyperparameter yourself.

3.  `autoselection`: Redeconve will use a procedure to select the optimal hyperparameter. In this procedure, a series of hyperparameter will be set in the vicinity of the hyperparameter selected by mode `default`, and Redeconve will use these hyperparameters to perform deconvolution separately, then return the result with the best hyperparameter. You can see Methods for details about how we determine the best hyperparameter.

Note that in this procedure, several rounds of deconvolution will be performed, so it may take a long time. Under such circumstances, parallel computing will be beneficial.

### 4. Parallel computing

Sometimes the reference will contain tens of thousands of cells, or the spatial transcriptomics will contain tens of thousands of spots (e.g. when the data is from Slide-seq), then parallel computing is useful. Redeconve uses the package "doSNOW" to achieve parallel computing (which means there is a progress bar). Related parameters are `dopar` and `ncores`.

-   `dopar` determines whether to use parallel computing or not.
-   `ncores` indicates the number of cores to be used in parallel computing. It's recommended to manually set this parameter rather than use the function `detectCores` to avoid underlying errors.

[! Important tips for parallel computing: !]{style="color:red"}

1.  Our underlying algorithm makes use of OpenBLAS, which may include parallel computing inside. Therefore, setting `system("export OPENBLAS_NUM_THREADS=1")` is necessary to avoid underlying errors.
2.  An error may be reported when the number of threads is too large : `Error in socketAccept(socket = socket, blocking = TRUE, open = "a+b",: all connections are in use`. If such error occurs, please reduce the number of cores.

### 5. Writing real-time results

Even with parallel computing, some dataset is still time-consuming. Redeconve is able to write results into disk in real time at the cost of some running speed. Related parameters are `realtime` and `dir`.

-   `realtime` determines whethers to write the results into disk in real time or not.
-   `dir` indicates the directory to write the results.

For real time results, the result of each spot will be write into a separate csv file, whose name is the barcode of the spot.

### 6. Other parameters

The left parameters are `cellnames`, `normalize` and `thre`.

-   `cellnames`: Chances are that you may not want to use all cells in reference to run deconvolution. Then you can indicate which cells will be used by this parameter. If you do not specify this parameter, all cells will be used.
-   `normalize`: Redeconve can also be used for bulk RNA-seq deconvolution. When doing this, normalization for reference is not required. When deconvoluting spatial transcriptomics, normalization is recommended.
-   `thre`: The estimated cell abundance will not be exactly 0. This parameter indicates that the abundance less than this value will be treated as 0. Generally this value does not need to be adjusted, and the result will remain the same within a relatively big range of this value.

### A demo

Next we will use a demo to give an example of how to use this function.

```{r}
## load the data
data(basic)

promises::promise_resolve(sc)
promises::promise_resolve(st)

## check the dimensions of sc and st
dim(sc)
dim(st)
```

```{r}
## check the number cells in each cell type
table(annotations[,2])
```

```{r,e}
## deconvolution
# this may take a long time
res = deconvoluting(sc,st,genemode="def",hpmode="def",dopar=T,ncores=8)
```

-   `sc` and `st` are separately reference and spatial transcriptomics.
-   `genemode` is set to `"default"` to guarantee accuracy.
-   `hpmode` is set to `"default"` to improve efficiency.
-   `dopar` is set to `TRUE` (default value) and `ncores` is set to 8. You can raise the number of cores to improve efficiency.
-   For this dataset is not very large, `realtime` is set to `FALSE` (default value).
-   We want to use all cells in deconvolution, so we do not need to specify `cellnames`. Also, we do not need to adjust `thre`.

------------------------------------------------------------------------

## Dealing with large scRNA datasets

Nowadays the scale of scRNA datasets are getting increasingly larger. Sometimes the reference cells can reach to sub-million or million level. Although Redeconve possesses high efficiency, directly using so much cells as reference is both time consuming and low yield. Here we provide two ways to deal with such situation:

### 1. Cell-type deconvolution

Like other methods, Redeconve can also perform deconvolution at cell-type level. This part shows how to do so.

```{r}
## get reference
ref = get.ref(sc,annotations,dopar = F)
```

```{r,message=FALSE}
## deconvolution
res.ct = deconvoluting(ref,st,genemode="def",hpmode="auto",dopar=T,ncores=8)
```

You can see that actually, only one more step is required to convert single-cell expression profile to that of cell type. The function `get.ref` will take the average expression of all cells in one cell type as the profile of that cell type. Note that `get.ref` also has parameters `dopar` and `ncores`, with which you can perform parallel computing when the scale of `sc` is too large.

For the main function, every thing is the same except that we set `hpmode` as `"autoselection"`, for there are only tens of cell types, then the speed is fast enough for us to run several rounds of deconvolution.

### 2. Sampling

Down-sampling is another way to deal with large scRNA datasets. Function `cell.sampling` is designed for this.

```{r}
idx = cell.sampling(ncells = 500, annotations, size = 200, prot=T)
sc.ds = sc[,idx[,1]]
```

This function returns a 2-columned matrix showing the barcodes and annotations of selected cells. The first parameter `ncells` is the original number of cells in scRNA dataset and the third `size` is the number of cells you want. The fourth parameter `prot` is to guarantee that at least one cell of each cell type in `annotations` is selected when set to `TRUE`.

```{r,message=FALSE}
res.ds = deconvoluting(sc.ds,st,genemode="def",hpmode="def",dopar=T,ncores=8)
```

------------------------------------------------------------------------


## Seurat interface

This package provided two functions to extract information from Seurat object to satisfy Redeconve's input:

```{r}
library(Seurat)
library(SeuratData)

## installing data
InstallData("pbmc3k")
InstallData("stxBrain")
```

```{r}
## sc
pbmc3k = UpdateSeuratObject(pbmc3k)
sclist = extract.seurat.sc(pbmc3k,anno_slot="seurat_annotations")

## st
stxBrain = LoadData('stxBrain', type = 'anterior1')
stlist = extract.seurat.st(stxBrain,coord_slot="anterior1")

## just for example
# res = deconvoluting(sclist[["expr"]],stlist[["expr"],genemode="filt",hpmode="auto",dopar=T,ncores=8)
```

------------------------------------------------------------------------

## Downstream analysis and visualization

Redeconve offers many built-in functions for downstream analysis and visualization.

### Gaining interpretability

For we included a normalization procedure, the result of `deconvoluting` has only relative significance. Here are some functions related:

1.  `to.proportion`: This function converts the result to proportion, i.e., the sum of cell abundance per spot is 1. This is convenient for visualization.

```{r}
res.prop = to.proportion(res)
```

2.  `to.absolute.abundance`: This function converts the result to absolute cell abundance with some priori knowledge. Parameter `aver.cell` is required to estimate the absolute abundance of each cell state. This value indicates the average number of cells in one spot. Users can determine it according to the platform of st data. For example, 10x Visium has a spot radius of about 50$\mu$m, so this value for a 10x Visium dataset is about 10.

```{r}
res.ab = to.absolute.abundance(res,aver.cell=10)
```

3.  `sc2type`: This function merges single-cell result to cell-type result according to annotations. Note that this function itself does not provide interpretability.

```{r}
res.ctmerge = sc2type(res,annotations)
```

### Cell occurrence

The function `cell.occur` shows the number of cells actually used in deconvolution, and the number of spots that every used cell occurs.

```{r}
cell.occur(res)
```

### Find cells of interest

Redeconve offers a function to plot mean vs sd, and highlighting those outliers. Users can then explore those cells. Note that this function can only be run after `cell.occur`, because it needs `occurred.cells` as input, which is an output of `cell.occur`.

```{r}
show.cellsofinterest(res, occurred.cells)
```

### Spatial distributiom of all or some specific cells

The function `spatial.cell.number` can plot the abundance of all cells (or cell types) or some cells you indicate.

```{r}
spatial.cell.number(res.ctmerge,cell.names="T cells & NK cells",coords=coords,size=4,pdfout=F)
```

Here we plotted the abundance of T cells & NK cells. When you want to plot the distribution of more than one cells, to set `pdfout` as `TRUE` is recommended.

### Spatial pie chart

The function `spatial.pie` can plot the spatial pie chart, showing the proportion of each cell type in each spatial spot.

```{r}
# For this is a demo, we use default colors
spatial.piechart(res.ctmerge,coords)
```

### Spot pie chart

The function `spot.pie` can plot the proportion of each cell type within a certain spot.

```{r}
spot.pie(res[,"X27x18"],title="X27x18")
```

### Co-localization

Function `coloc.network`infers a co-localization network from correlation of abundance of cells or cell types.

```{r}
corr = cor(t(res.ctmerge),method = "pearson")
celltypes = levels(as.factor(annotations[,2]))
g = coloc.network(corr,thresh=0.4,cell.type=T,annotations=celltypes,ntypes=length(celltypes))
```

The parameter `thresh` means that only correlation coefficient larger than this value is regarded as an edge.


```{r}
# plotting with igraph
plot(g, vertex.size = 20, vertex.label.cex = 0.7, layout=layout_in_circle)
```

<!-- ### Spatial expression profile visulization and imputation -->

<!-- Function `spatial.gene` is similar to `spatial.cell.number`, the difference is that this function plot the spatial expression of some genes. -->

<!-- ```{r} -->
<!-- library(patchwork) -->

<!-- p = spatial.gene(st,coords,gene.list = c("CD3D","CD3E","CD3G","CD8A"),pdf=F) -->
<!-- p[[1]]+p[[2]]+p[[3]]+p[[4]]+plot_layout(ncol=2) -->
<!-- ``` -->

<!-- We can see that some CD8+ spots are not CD3+, which is unreasonable. This is because of spatial "drop-outs", in which some genes are not corrected detected.  -->

<!-- Redeconve can reconstruct the spatial expression profile by the estimated absolute cell abundance and the single-cell expression profile, thus enabling imputation of the drop-outs. -->

<!-- ```{r} -->
<!-- rec = profile.imputation(res.ab, sc) -->
<!-- ## note that absolute cell abundance is required. -->
<!-- ``` -->

<!-- ```{r} -->
<!-- p_rec = spatial.gene(rec,coords,gene.list = c("CD3D","CD3E","CD3G","CD8A"),pdf=T) -->
<!-- p_rec[[1]]+p_rec[[2]]+p_rec[[3]]+p_rec[[4]]+plot_layout(ncol=2) -->
<!-- ``` -->
